Race in `TestServerRegisterPeer`

```
2023-04-13T15:51:08.5219927Z === RUN   TestServerRegisterPeer
2023-04-13T15:51:08.5220513Z     logger.go:130: 2023-04-13T15:45:33.453Z	INFO	ExtensiblePoolSize is not set or wrong, using default value	{"ExtensiblePoolSize": 20}
2023-04-13T15:51:08.5221066Z     logger.go:130: 2023-04-13T15:45:33.453Z	INFO	bad AttemptConnPeers configured, using the default value	{"configured": 0, "actual": 20}
2023-04-13T15:51:08.5221462Z     logger.go:130: 2023-04-13T15:45:33.454Z	INFO	node started	{"blockHeight": 0, "headerHeight": 0}
2023-04-13T15:51:08.5221852Z     logger.go:130: 2023-04-13T15:45:33.454Z	INFO	node reached synchronized state, starting services
2023-04-13T15:51:08.5222248Z     logger.go:130: 2023-04-13T15:45:33.454Z	INFO	new peer connected	{"addr": "0.0.0.0:1", "peerCount": 1}
2023-04-13T15:51:08.5222332Z ==================
2023-04-13T15:51:08.5222424Z WARNING: DATA RACE
2023-04-13T15:51:08.5222551Z Read at 0x00c0004c6ba3 by goroutine 79:
2023-04-13T15:51:08.5222681Z   testing.(*common).logDepth()
2023-04-13T15:51:08.5222993Z       /opt/hostedtoolcache/go/1.18.10/x64/src/testing/testing.go:882 +0x4ce
2023-04-13T15:51:08.5223112Z   testing.(*common).log()
2023-04-13T15:51:08.5223413Z       /opt/hostedtoolcache/go/1.18.10/x64/src/testing/testing.go:869 +0x9c
2023-04-13T15:51:08.5223529Z   testing.(*common).Logf()
2023-04-13T15:51:08.5223814Z       /opt/hostedtoolcache/go/1.18.10/x64/src/testing/testing.go:920 +0x67
2023-04-13T15:51:08.5223920Z   testing.(*T).Logf()
2023-04-13T15:51:08.5224110Z       <autogenerated>:1 +0x75
2023-04-13T15:51:08.5224294Z   go.uber.org/zap/zaptest.testingWriter.Write()
2023-04-13T15:51:08.5224633Z       /home/runner/go/pkg/mod/go.uber.org/zap@v1.24.0/zaptest/logger.go:130 +0x12c
2023-04-13T15:51:08.5224797Z   go.uber.org/zap/zaptest.(*testingWriter).Write()
2023-04-13T15:51:08.5224941Z       <autogenerated>:1 +0x7e
2023-04-13T15:51:08.5225094Z   go.uber.org/zap/zapcore.(*ioCore).Write()
2023-04-13T15:51:08.5225463Z       /home/runner/go/pkg/mod/go.uber.org/zap@v1.24.0/zapcore/core.go:99 +0x196
2023-04-13T15:51:08.5225638Z   go.uber.org/zap/zapcore.(*CheckedEntry).Write()
2023-04-13T15:51:08.5225962Z       /home/runner/go/pkg/mod/go.uber.org/zap@v1.24.0/zapcore/entry.go:255 +0x2cc
2023-04-13T15:51:08.5226096Z   go.uber.org/zap.(*Logger).Warn()
2023-04-13T15:51:08.5226397Z       /home/runner/go/pkg/mod/go.uber.org/zap@v1.24.0/logger.go:228 +0x6d
2023-04-13T15:51:08.5226656Z   github.com/nspcc-dev/neo-go/pkg/network.(*Server).run()
2023-04-13T15:51:08.5227039Z       /home/runner/work/neo-go/neo-go/pkg/network/server.go:490 +0x12e8
2023-04-13T15:51:08.5227302Z   github.com/nspcc-dev/neo-go/pkg/network.(*Server).Start()
2023-04-13T15:51:08.5227673Z       /home/runner/work/neo-go/neo-go/pkg/network/server.go:288 +0x646
2023-04-13T15:51:08.5228024Z   github.com/nspcc-dev/neo-go/pkg/network.TestServerStartAndShutdown.func1.2()
2023-04-13T15:51:08.5228401Z       /home/runner/work/neo-go/neo-go/pkg/network/server_test.go:94 +0x39
2023-04-13T15:51:08.5228418Z 
2023-04-13T15:51:08.5228565Z Previous write at 0x00c0004c6ba3 by goroutine 42:
2023-04-13T15:51:08.5228689Z   testing.tRunner.func1()
2023-04-13T15:51:08.5228994Z       /opt/hostedtoolcache/go/1.18.10/x64/src/testing/testing.go:1426 +0x7af
2023-04-13T15:51:08.5229115Z   runtime.deferreturn()
2023-04-13T15:51:08.5229410Z       /opt/hostedtoolcache/go/1.18.10/x64/src/runtime/panic.go:436 +0x32
2023-04-13T15:51:08.5229523Z   testing.(*T).Run.func1()
2023-04-13T15:51:08.5229817Z       /opt/hostedtoolcache/go/1.18.10/x64/src/testing/testing.go:1486 +0x47
2023-04-13T15:51:08.5229829Z 
2023-04-13T15:51:08.5229946Z Goroutine 79 (running) created at:
2023-04-13T15:51:08.5230287Z   github.com/nspcc-dev/neo-go/pkg/network.TestServerStartAndShutdown.func1()
2023-04-13T15:51:08.5230667Z       /home/runner/work/neo-go/neo-go/pkg/network/server_test.go:94 +0xf1
2023-04-13T15:51:08.5230771Z   testing.tRunner()
2023-04-13T15:51:08.5231076Z       /opt/hostedtoolcache/go/1.18.10/x64/src/testing/testing.go:1439 +0x213
2023-04-13T15:51:08.5231189Z   testing.(*T).Run.func1()
2023-04-13T15:51:08.5231527Z       /opt/hostedtoolcache/go/1.18.10/x64/src/testing/testing.go:1486 +0x47
2023-04-13T15:51:08.5231575Z 
2023-04-13T15:51:08.5231695Z Goroutine 42 (finished) created at:
2023-04-13T15:51:08.5231797Z   testing.(*T).Run()
2023-04-13T15:51:08.5232087Z       /opt/hostedtoolcache/go/1.18.10/x64/src/testing/testing.go:1486 +0x724
2023-04-13T15:51:08.5232200Z   testing.runTests.func1()
2023-04-13T15:51:08.5232496Z       /opt/hostedtoolcache/go/1.18.10/x64/src/testing/testing.go:1839 +0x99
2023-04-13T15:51:08.5232601Z   testing.tRunner()
2023-04-13T15:51:08.5232891Z       /opt/hostedtoolcache/go/1.18.10/x64/src/testing/testing.go:1439 +0x213
2023-04-13T15:51:08.5232994Z   testing.runTests()
2023-04-13T15:51:08.5233290Z       /opt/hostedtoolcache/go/1.18.10/x64/src/testing/testing.go:1837 +0x7e4
2023-04-13T15:51:08.5233389Z   testing.(*M).Run()
2023-04-13T15:51:08.5233673Z       /opt/hostedtoolcache/go/1.18.10/x64/src/testing/testing.go:1719 +0xa71
2023-04-13T15:51:08.5233765Z   main.main()
2023-04-13T15:51:08.5233930Z       _testmain.go:139 +0x2e4
2023-04-13T15:51:08.5234016Z ==================
2023-04-13T15:51:08.5234526Z     logger.go:130: 2023-04-13T15:45:33.475Z	INFO	started protocol	{"addr": "0.0.0.0:1", "userAgent": "fake", "startHeight": 0, "id": 0}
2023-04-13T15:51:08.5235092Z     logger.go:130: 2023-04-13T15:45:33.475Z	INFO	new peer connected	{"addr": "0.0.0.0:2", "peerCount": 2}
2023-04-13T15:51:08.5235592Z     logger.go:130: 2023-04-13T15:45:33.475Z	INFO	started protocol	{"addr": "0.0.0.0:2", "userAgent": "fake", "startHeight": 0, "id": 1}
2023-04-13T15:51:08.5236062Z     logger.go:130: 2023-04-13T15:45:33.491Z	INFO	new peer connected	{"addr": "0.0.0.0:3", "peerCount": 3}
2023-04-13T15:51:08.5236177Z peer dropped: max peers reached
2023-04-13T15:51:08.5236661Z     logger.go:130: 2023-04-13T15:45:33.491Z	WARN	peer disconnected	{"addr": "0.0.0.0:2", "error": "max peers reached", "peerCount": 2}
2023-04-13T15:51:08.5237212Z     logger.go:130: 2023-04-13T15:45:33.599Z	WARN	peer disconnected	{"addr": "0.0.0.0:3", "error": "identical node id", "peerCount": 1}
2023-04-13T15:51:08.5237575Z     logger.go:130: 2023-04-13T15:45:33.649Z	INFO	shutting down server	{"peers": 1}
2023-04-13T15:51:08.5237692Z peer dropped: server shutdown
2023-04-13T15:51:08.5237927Z     testing.go:1312: race detected during execution of test
2023-04-13T15:51:08.5238128Z --- FAIL: TestServerRegisterPeer (0.20s)
```

--- a/cli/server/server.go
+++ b/cli/server/server.go
@@ -498,7 +498,7 @@ func startServer(ctx *cli.Context) error {
 	rpcServer := rpcsrv.New(chain, cfg.ApplicationConfiguration.RPC, serv, oracleSrv, log, errChan)
 	serv.AddService(&rpcServer)
 
-	go serv.Start()
+	serv.Start()
 	if !cfg.ApplicationConfiguration.RPC.StartWhenSynchronized {
 		// Run RPC server in a separate routine. This is necessary to avoid a potential
 		// deadlock: Start() can write errors to errChan which is not yet read in the

--- a/internal/testcli/executor.go
+++ b/internal/testcli/executor.go
@@ -164,7 +164,7 @@ func NewTestChain(t *testing.T, f func(*config.Config), run bool) (*core.Blockch
 	})
 	require.NoError(t, err)
 	netSrv.AddConsensusService(cons, cons.OnPayload, cons.OnTransaction)
-	go netSrv.Start()
+	netSrv.Start()
 	errCh := make(chan error, 2)
 	rpcServer := rpcsrv.New(chain, cfg.ApplicationConfiguration.RPC, netSrv, nil, logger, errCh)
 	rpcServer.Start()

--- a/pkg/consensus/consensus.go
+++ b/pkg/consensus/consensus.go
@@ -295,6 +295,7 @@ func (s *service) Shutdown() {
 			s.wallet.Close()
 		}
 	}
+	_ = s.log.Sync()
 }
 
 func (s *service) eventLoop() {

--- a/pkg/consensus/consensus.go
+++ b/pkg/consensus/consensus.go
@@ -281,7 +281,6 @@ func (s *service) Start() {
 		b, _ := s.Chain.GetBlock(s.Chain.CurrentBlockHash()) // Can't fail, we have some current block!
 		s.lastTimestamp = b.Timestamp
 		s.dbft.Start(s.lastTimestamp * nsInMs)
-		s.Chain.SubscribeForBlocks(s.blockEvents)
 		go s.eventLoop()
 	}
 }
@@ -299,6 +298,14 @@ func (s *service) Shutdown() {
 }
 
 func (s *service) eventLoop() {
+	s.Chain.SubscribeForBlocks(s.blockEvents)
+
+	// Manually sync up with potentially missed fresh blocks that may be added by blockchain
+	// before the subscription.
+	b, _ := s.Chain.GetBlock(s.Chain.CurrentBlockHash()) // Can't fail, we have some current block!
+	if b.Timestamp >= s.lastTimestamp {
+		s.handleChainBlock(b)
+	}
 events:
 	for {
 		select {

--- a/pkg/core/blockchain.go
+++ b/pkg/core/blockchain.go
@@ -1447,6 +1447,7 @@ func (bc *Blockchain) Close() {
 	close(bc.stopCh)
 	<-bc.runToExitCh
 	bc.addLock.Unlock()
+	_ = bc.log.Sync()
 }
 
 // AddBlock accepts successive block for the Blockchain, verifies it and

--- a/pkg/network/server.go
+++ b/pkg/network/server.go
@@ -320,6 +320,8 @@ func (s *Server) Shutdown() {
 	}
 	close(s.quit)
 	<-s.relayFin
+
+	_ = s.log.Sync()
 }
 
 // AddService allows to add a service to be started/stopped by Server.

--- a/pkg/network/server.go
+++ b/pkg/network/server.go
@@ -138,6 +138,9 @@ type (
 		stateSync StateSync
 
 		log *zap.Logger
+
+		// started used to Start and Shutdown server only once.
+		started atomic.Bool
 	}
 
 	peerDrop struct {
@@ -262,8 +265,12 @@ func (s *Server) ID() uint32 {
 }
 
 // Start will start the server and its underlying transport. Calling it twice
-// is an error. Caller should wait for Start to finish for normal server operation.
+// is a no-op. Caller should wait for Start to finish for normal server operation.
 func (s *Server) Start() {
+	if !s.started.CompareAndSwap(false, true) {
+		s.log.Info("node server already started")
+		return
+	}
 	s.log.Info("node started",
 		zap.Uint32("blockHeight", s.chain.BlockHeight()),
 		zap.Uint32("headerHeight", s.chain.HeaderHeight()))
@@ -288,9 +295,12 @@ func (s *Server) Start() {
 	go s.run()
 }
 
-// Shutdown disconnects all peers and stops listening. Calling it twice is an error,
+// Shutdown disconnects all peers and stops listening. Calling it twice is a no-op,
 // once stopped the same intance of the Server can't be started again by calling Start.
 func (s *Server) Shutdown() {
+	if !s.started.CompareAndSwap(true, false) {
+		return
+	}
 	s.log.Info("shutting down server", zap.Int("peers", s.PeerCount()))
 	for _, tr := range s.transports {
 		tr.Close()

--- a/pkg/network/server.go
+++ b/pkg/network/server.go
@@ -296,7 +296,7 @@ func (s *Server) Start() {
 }
 
 // Shutdown disconnects all peers and stops listening. Calling it twice is a no-op,
-// once stopped the same intance of the Server can't be started again by calling Start.
+// once stopped the same instance of the Server can't be started again by calling Start.
 func (s *Server) Shutdown() {
 	if !s.started.CompareAndSwap(true, false) {
 		return

--- a/pkg/network/server.go
+++ b/pkg/network/server.go
@@ -124,12 +124,14 @@ type (
 		lastRequestedBlock atomic.Uint32
 		// lastRequestedHeader contains a height of the last requested header.
 		lastRequestedHeader atomic.Uint32
-
-		register   chan Peer
-		unregister chan peerDrop
-		handshake  chan Peer
-		quit       chan struct{}
-		relayFin   chan struct{}
+		register            chan Peer
+		unregister          chan peerDrop
+		handshake           chan Peer
+		quit                chan struct{}
+		relayFin            chan struct{}
+		runFin              chan struct{}
+		broadcastTxFin      chan struct{}
+		runProtoFin         chan struct{}
 
 		transactions chan *transaction.Transaction
 
@@ -141,6 +143,8 @@ type (
 
 		// started used to Start and Shutdown server only once.
 		started atomic.Bool
+
+		txHandlerLoopWG sync.WaitGroup
 	}
 
 	peerDrop struct {
@@ -183,6 +187,9 @@ func newServerFromConstructors(config ServerConfig, chain Ledger, stSync StateSy
 		config:         chain.GetConfig().ProtocolConfiguration,
 		quit:           make(chan struct{}),
 		relayFin:       make(chan struct{}),
+		runFin:         make(chan struct{}),
+		broadcastTxFin: make(chan struct{}),
+		runProtoFin:    make(chan struct{}),
 		register:       make(chan Peer),
 		unregister:     make(chan peerDrop),
 		handshake:      make(chan Peer),
@@ -279,6 +286,7 @@ func (s *Server) Start() {
 	s.initStaleMemPools()
 
 	var txThreads = optimalNumOfThreads()
+	s.txHandlerLoopWG.Add(txThreads)
 	for i := 0; i < txThreads; i++ {
 		go s.txHandlerLoop()
 	}
@@ -319,7 +327,11 @@ func (s *Server) Shutdown() {
 		s.notaryRequestPool.StopSubscriptions()
 	}
 	close(s.quit)
+	<-s.broadcastTxFin
+	<-s.runProtoFin
 	<-s.relayFin
+	<-s.runFin
+	s.txHandlerLoopWG.Wait()
 
 	_ = s.log.Sync()
 }
@@ -435,6 +447,7 @@ func (s *Server) run() {
 		addrTimer        = time.NewTimer(peerCheckTime)
 		peerTimer        = time.NewTimer(s.ProtoTickInterval)
 	)
+	defer close(s.runFin)
 	defer addrTimer.Stop()
 	defer peerTimer.Stop()
 	go s.runProto()
@@ -533,6 +546,7 @@ func (s *Server) run() {
 
 // runProto is a goroutine that manages server-wide protocol events.
 func (s *Server) runProto() {
+	defer close(s.runProtoFin)
 	pingTimer := time.NewTimer(s.PingInterval)
 	for {
 		prevHeight := s.chain.BlockHeight()
@@ -1135,6 +1149,7 @@ func (s *Server) handleTxCmd(tx *transaction.Transaction) error {
 }
 
 func (s *Server) txHandlerLoop() {
+	defer s.txHandlerLoopWG.Done()
 txloop:
 	for {
 		select {
@@ -1651,6 +1666,7 @@ func (s *Server) broadcastTxLoop() {
 		batchSize = 42
 	)
 
+	defer close(s.broadcastTxFin)
 	txs := make([]util.Uint256, 0, batchSize)
 	var timer *time.Timer
 

--- a/pkg/network/server.go
+++ b/pkg/network/server.go
@@ -262,7 +262,7 @@ func (s *Server) ID() uint32 {
 }
 
 // Start will start the server and its underlying transport. Calling it twice
-// is an error.
+// is an error. Caller should wait for Start to finish for normal server operation.
 func (s *Server) Start() {
 	s.log.Info("node started",
 		zap.Uint32("blockHeight", s.chain.BlockHeight()),
@@ -285,7 +285,7 @@ func (s *Server) Start() {
 	setServerAndNodeVersions(s.UserAgent, strconv.FormatUint(uint64(s.id), 10))
 	setNeoGoVersion(config.Version)
 	setSeverID(strconv.FormatUint(uint64(s.id), 10))
-	s.run()
+	go s.run()
 }
 
 // Shutdown disconnects all peers and stops listening. Calling it twice is an error,

--- a/pkg/network/server_test.go
+++ b/pkg/network/server_test.go
@@ -96,10 +96,12 @@ func TestServerStartAndShutdown(t *testing.T) {
 		require.Eventually(t, func() bool { return 1 == s.PeerCount() }, time.Second, time.Millisecond*10)
 
 		assert.True(t, s.transports[0].(*fakeTransp).started.Load())
+		require.True(t, s.started.Load())
 		assert.Nil(t, s.txCallback)
 
 		s.Shutdown()
 
+		require.False(t, s.started.Load())
 		require.True(t, s.transports[0].(*fakeTransp).closed.Load())
 		err, ok := p.droppedWith.Load().(error)
 		require.True(t, ok)
@@ -115,10 +117,33 @@ func TestServerStartAndShutdown(t *testing.T) {
 		s.register <- p
 
 		assert.True(t, s.services["fake"].(*fakeConsensus).started.Load())
+		require.True(t, s.started.Load())
 
 		s.Shutdown()
 
+		require.False(t, s.started.Load())
 		require.True(t, s.services["fake"].(*fakeConsensus).stopped.Load())
+	})
+	t.Run("double start", func(t *testing.T) {
+		s := newTestServer(t, ServerConfig{})
+		startWithCleanup(t, s)
+
+		// Attempt to start the server again.
+		s.Start()
+
+		require.True(t, s.started.Load(), "server should still be marked as started after second Start call")
+	})
+	t.Run("double shutdown", func(t *testing.T) {
+		s := newTestServer(t, ServerConfig{})
+		s.Start()
+		require.True(t, s.started.Load(), "server should still be marked as started after second Start call")
+		s.Shutdown()
+
+		require.False(t, s.started.Load(), "server should be marked as not started after second Shutdown call")
+		// Attempt to shutdown the server again.
+		s.Shutdown()
+		// Verify the server state remains unchanged and is still considered shutdown.
+		require.False(t, s.started.Load(), "server should remain shutdown after second call")
 	})
 }
 

--- a/pkg/network/server_test.go
+++ b/pkg/network/server_test.go
@@ -90,7 +90,7 @@ func TestServerStartAndShutdown(t *testing.T) {
 	t.Run("no consensus", func(t *testing.T) {
 		s := newTestServer(t, ServerConfig{})
 
-		go s.Start()
+		s.Start()
 		p := newLocalPeer(t, s)
 		s.register <- p
 		require.Eventually(t, func() bool { return 1 == s.PeerCount() }, time.Second, time.Millisecond*10)
@@ -110,7 +110,7 @@ func TestServerStartAndShutdown(t *testing.T) {
 		cons := new(fakeConsensus)
 		s.AddConsensusService(cons, cons.OnPayload, cons.OnTransaction)
 
-		go s.Start()
+		s.Start()
 		p := newLocalPeer(t, s)
 		s.register <- p
 
@@ -312,7 +312,7 @@ func TestServerNotSendsVerack(t *testing.T) {
 	s.id = 1
 	finished := make(chan struct{})
 	go func() {
-		s.run()
+		go s.run()
 		close(finished)
 	}()
 	t.Cleanup(func() {
@@ -389,7 +389,7 @@ func startTestServer(t *testing.T, protocolCfg ...func(*config.Blockchain)) *Ser
 }
 
 func startWithCleanup(t *testing.T, s *Server) {
-	go s.Start()
+	s.Start()
 	t.Cleanup(func() {
 		s.Shutdown()
 	})

--- a/pkg/services/metrics/metrics.go
+++ b/pkg/services/metrics/metrics.go
@@ -75,4 +75,5 @@ func (ms *Service) ShutDown() {
 			ms.log.Error("can't shut service down", zap.String("endpoint", srv.Addr), zap.Error(err))
 		}
 	}
+	_ = ms.log.Sync()
 }

--- a/pkg/services/notary/notary.go
+++ b/pkg/services/notary/notary.go
@@ -175,13 +175,13 @@ func (n *Notary) Start() {
 		return
 	}
 	n.Config.Log.Info("starting notary service")
-	n.Config.Chain.SubscribeForBlocks(n.blocksCh)
-	n.mp.SubscribeForTransactions(n.reqCh)
 	go n.newTxCallbackLoop()
 	go n.mainLoop()
 }
 
 func (n *Notary) mainLoop() {
+	n.Config.Chain.SubscribeForBlocks(n.blocksCh)
+	n.mp.SubscribeForTransactions(n.reqCh)
 mainloop:
 	for {
 		select {

--- a/pkg/services/notary/notary.go
+++ b/pkg/services/notary/notary.go
@@ -228,6 +228,7 @@ func (n *Notary) Shutdown() {
 	close(n.stopCh)
 	<-n.done
 	n.wallet.Close()
+	_ = n.Config.Log.Sync()
 }
 
 // IsAuthorized returns whether Notary service currently is authorized to collect

--- a/pkg/services/oracle/oracle.go
+++ b/pkg/services/oracle/oracle.go
@@ -192,6 +192,7 @@ func (o *Oracle) Shutdown() {
 	o.ResponseHandler.Shutdown()
 	<-o.done
 	o.wallet.Close()
+	_ = o.Log.Sync()
 }
 
 // Start runs the oracle service in a separate goroutine.

--- a/pkg/services/rpcsrv/server.go
+++ b/pkg/services/rpcsrv/server.go
@@ -482,6 +482,7 @@ func (s *Server) Shutdown() {
 
 	// Wait for handleSubEvents to finish.
 	<-s.subEventsToExitCh
+	_ = s.log.Sync()
 }
 
 // SetOracleHandler allows to update oracle handler used by the Server.

--- a/pkg/services/rpcsrv/subscription_test.go
+++ b/pkg/services/rpcsrv/subscription_test.go
@@ -99,7 +99,7 @@ func TestSubscriptions(t *testing.T) {
 	defer chain.Close()
 	defer rpcSrv.Shutdown()
 
-	go rpcSrv.coreServer.Start()
+	rpcSrv.coreServer.Start()
 	defer rpcSrv.coreServer.Shutdown()
 
 	for _, feed := range subFeeds {
@@ -395,7 +395,7 @@ func TestFilteredNotaryRequestSubscriptions(t *testing.T) {
 	}
 
 	chain, rpcSrv, c, respMsgs, finishedFlag := initCleanServerAndWSClient(t)
-	go rpcSrv.coreServer.Start()
+	rpcSrv.coreServer.Start()
 
 	defer chain.Close()
 	defer rpcSrv.Shutdown()

--- a/pkg/services/stateroot/validators.go
+++ b/pkg/services/stateroot/validators.go
@@ -77,6 +77,7 @@ func (s *service) Shutdown() {
 	if s.wallet != nil {
 		s.wallet.Close()
 	}
+	_ = s.log.Sync()
 }
 
 func (s *service) signAndSend(r *state.MPTRoot) error {

--- a/pkg/services/stateroot/validators.go
+++ b/pkg/services/stateroot/validators.go
@@ -29,11 +29,11 @@ func (s *service) Start() {
 		return
 	}
 	s.log.Info("starting state validation service")
-	s.chain.SubscribeForBlocks(s.blockCh)
 	go s.run()
 }
 
 func (s *service) run() {
+	s.chain.SubscribeForBlocks(s.blockCh)
 runloop:
 	for {
 		select {

